src/container_of.rs

```diff
@@ -0,0 +1,34 @@
+/// Calculates the address of a containing struct from a pointer to one of its
+/// fields.
+///
+/// # Safety
+///
+/// This is unsafe because it assumes that the given expression is a valid
+/// pointer to the specified field of some container type.
+///
+/// ## Examples
+/// ```
+/// #[macro_use]
+/// extern crate memoffset;
+///
+/// #[repr(C, packed)]
+/// struct Foo {
+///     a: u32,
+///     b: u64,
+///     c: [u8; 5]
+/// }
+///
+/// fn main() {
+///     let container = Foo { a: 1, b: 2, c: [3; 5] };
+///     let field = &container.b;
+///     let container2: *const Foo = unsafe { container_of!(field, Foo, b) };
+///     assert_eq!(&container as *const Foo, container2);
+/// }
+/// ```
+#[macro_export(local_inner_macros)]
+macro_rules! container_of {
+    ($ptr:expr, $container:path, $field:tt) => {
+        ($ptr as *const _ as *const u8).offset(-(offset_of!($container, $field) as isize))
+            as *mut $container
```

Comment:
You are starting with a `*const` and then turning it into a `*mut` here, that's an awful foot-gun. Please keep mutability consistent.

Reply:
Doing it this way allows the macro to be used with both `*const` and `*mut`, since `*mut` coerces to `*const` but the reverse is not true. There shouldn't be any safety issues since `*const` and `*mut` are both equivalent to `usize`.

Reply:
See https://github.com/Gilnaa/memoffset/issues/24: while the *types* are equivalent, the question of whether the *casts* are equivalent is still wide open.

`&mut x as *const _` produces a read-only pointer, according to current Stacked Borrows. Fixing this, if we consider it a bug, requires a significant extension of the machinery.

So, please don't cast references to `*const` and then mutate that later.

Reply:
I've changed the macro to always return `*const`. This is enough for my use case in `intrusive-collections` since the fields I manipulate are wrapped in `Cell` anyways.

Reply:
Thanks!

We can add a `mut` version later. Though probably we should do it sooner rather than later because people might just cast the pointer around otherwise. But it doesn't have to happen in this PR.